internal/controller/supervisorconfig/upstreamwatcher/ldap_upstream_watcher.go

```diff
@@ -0,0 +1,367 @@
+			// Now figure out which version of the bind Secret was used during that previous validation.
+			matches := secretVersionParser.FindStringSubmatch(c.Message)
+			if len(matches) != 2 {
+				continue
+			}
+			validatedSecretVersion := matches[1]
+			if validatedSecretVersion == currentSecretVersion {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func (c *ldapWatcherController) validTLSCondition(message string) *v1alpha1.Condition {
+	return &v1alpha1.Condition{
+		Type:    typeTLSConfigurationValid,
+		Status:  v1alpha1.ConditionTrue,
+		Reason:  reasonSuccess,
+		Message: message,
+	}
+}
+
+func (c *ldapWatcherController) invalidTLSCondition(message string) *v1alpha1.Condition {
+	return &v1alpha1.Condition{
+		Type:    typeTLSConfigurationValid,
+		Status:  v1alpha1.ConditionFalse,
+		Reason:  reasonInvalidTLSConfig,
+		Message: message,
+	}
+}
+
+func (c *ldapWatcherController) validateSecret(upstream *v1alpha1.LDAPIdentityProvider, config *upstreamldap.ProviderConfig) (*v1alpha1.Condition, string) {
+	secretName := upstream.Spec.Bind.SecretName
+
+	secret, err := c.secretInformer.Lister().Secrets(upstream.Namespace).Get(secretName)
+	if err != nil {
+		return &v1alpha1.Condition{
+			Type:    typeBindSecretValid,
+			Status:  v1alpha1.ConditionFalse,
+			Reason:  reasonNotFound,
+			Message: err.Error(),
+		}, ""
+	}
+
+	if secret.Type != corev1.SecretTypeBasicAuth {
+		return &v1alpha1.Condition{
+			Type:   typeBindSecretValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonWrongType,
+			Message: fmt.Sprintf("referenced Secret %q has wrong type %q (should be %q)",
+				secretName, secret.Type, corev1.SecretTypeBasicAuth),
+		}, secret.ResourceVersion
+	}
+
+	config.BindUsername = string(secret.Data[corev1.BasicAuthUsernameKey])
+	config.BindPassword = string(secret.Data[corev1.BasicAuthPasswordKey])
+	if len(config.BindUsername) == 0 || len(config.BindPassword) == 0 {
+		return &v1alpha1.Condition{
+			Type:   typeBindSecretValid,
+			Status: v1alpha1.ConditionFalse,
+			Reason: reasonMissingKeys,
+			Message: fmt.Sprintf("referenced Secret %q is missing required keys %q",
+				secretName, []string{corev1.BasicAuthUsernameKey, corev1.BasicAuthPasswordKey}),
+		}, secret.ResourceVersion
+	}
+
+	return &v1alpha1.Condition{
+		Type:    typeBindSecretValid,
+		Status:  v1alpha1.ConditionTrue,
+		Reason:  reasonSuccess,
+		Message: "loaded bind secret",
+	}, secret.ResourceVersion
+}
+
+func (c *ldapWatcherController) updateStatus(ctx context.Context, upstream *v1alpha1.LDAPIdentityProvider, conditions []*v1alpha1.Condition) bool {
+	log := klogr.New().WithValues("namespace", upstream.Namespace, "name", upstream.Name)
+	updated := upstream.DeepCopy()
+
+	hadErrorCondition := mergeConditions(conditions, upstream.Generation, &updated.Status.Conditions, log)
+
+	updated.Status.Phase = v1alpha1.LDAPPhaseReady
+	if hadErrorCondition {
+		updated.Status.Phase = v1alpha1.LDAPPhaseError
+	}
```

Comment:
We should delete the Pending condition if we are not using it.

Reply:
Pending is the default value in the CRD. Not sure which versions of K8s actually honors default values in CRDs though. This was copied from the corresponding OIDCIdentityProvider type to be consistent.